### ISSUE TITLE
Full codebase.

### ISSUE THREAD

**Issue by rozeappletree** [yc015/TalkTuner-chatbot-llm-dashboard#1]
Hi.

Amazing work on TalkTuner! I am trying to reproduce it as I write for my application process for [MATS under Neel Nanda](https://docs.google.com/document/d/1p-ggQV3vVWIQuCccXEl1fD0thJOgXimlbBpGk6FI32I/edit?tab=t.0).

The [github repo](https://github.com/yc015/TalkTuner-chatbot-llm-dashboard) mentions that the "repo is under construction".

It has already been 2years now, would you be kind enough to release the full codebase?

If releasing it is not possible, please let me know too.
Looking forward to hearing from you.

Rakesh.
https://github.com/rozeappletree
https://www.linkedin.com/in/asapanna-rakesh
https://asra.dev/

**Comment by rozeappletree** [comment]
ℹ️ Got reply from Martin, they are busy. Shared my findings. Adding here too.

[martin.pdf](https://github.com/user-attachments/files/31759784/martin.pdf)

**Comment by yc015** [comment]
Thank you for your interest in our work. I am copying my reply to your email below just for future reference.

"To my understanding, what you are mainly looking for is the web app code for TalkTuner? Please correct me if I am mistaken.

For the web app implementation, I need to clean up the code a bit before sharing it. I believe I hard-coded the server address, MongoDB passkey, a few etc. at one point. You can expect me to make it available by the end of this week or early next week, likely next Monday or Tuesday.

Meanwhile, if you need a working implementation urgently for your application, I would recommend taking a look at a great community implementation by Cat McGee::[ https://github.com/catmcgee/TalkTuner](https://github.com/catmcgee/TalkTuner), She forked our repository and recreated the web UI using our research code.

For your other questions, here are some quick responses.
> after extracting dataset zip file I noticed that there's no openai_age_3 or openai_socioeconomic_3 at all (only _1 and _2 exist for those)
        * the training code at  [train_read_and_controlling_probes.ipynb](https://github.com/rozeappletree/TalkTuner-chatbot-llm-dashboard/blob/main/notebooks/train_probes/train_read_and_controlling_probes.ipynb)   suggests that these folders do exists - either the training code is old or the data folder.

That probably means there simply is no _3 split for openai_age and openai_socioeconomic.

In fact, in the notebook you provided (thanks for putting that together, by the way), the total number of conversations in the zip file matches the number of conversations we report in Table 1 of the paper (note: you need to exclude the llama_gender_3 partition. It's an extra partition that I uploaded by accident). For the llama_gender_3 partition, we did not use that partition in the experiments. If you run the training notebook train_read_and_controlling_probes.ipynb, you will see that llama_gender_3 is not loaded for training. 

So, as far as I can tell, the dataset itself is complete.  

<img width="230" height="257" alt="Image" src="https://github.com/user-attachments/assets/fbfc8e40-5b0c-4490-acdd-aca00af816f7" />

<img width="410" height="76" alt="Image" src="https://github.com/user-attachments/assets/79433009-07c7-4384-98d3-49b0b7b9df73" />

In the paper, you have mentioned that only openai was used to generate synthetic data. Then why do you also have llama generated data?
In Section 4.1, we did mention that both GPT-3.5 and LLaMA2-Chat were used to synthesize the data: “Using the role-playing technique, we generated synthetic conversations using GPT-3.5 and LLaMa2Chat.”

> The code seems to exist for-show-purpose only- did not seem as if it was used for training

Thank you for catching the seeds typo. That variable was used in an earlier version of the experiment where we trained the same probes on multiple train/test splits to measure stability across splits.

My observation at the time was that the s.d. of test accuracy across different splits was very small, so I eventually removed it. The training notebook should be runnable after removing the leftover reference to seeds.

> In NVIDIA L40S, the training takes a huge amount of time unlike what is mentioned at the end of the paper: 3mins in A100... in L40S, it took me 9mins to generate one single probe (for layer 0) 

For layer 0, the code is doing more than just training the probe. It also performs a forward pass through the LLM to collect the hidden states needed for probe training. Because the conversations can be fairly long, this hidden-state collection can take a significant amount of time.

However, that forward pass collects hidden states for all layers at once. In other words, it should only need to happen once per attribute.  If you let the code continue running, you will see training the probes for layer 1 and subsequent layers are much faster. 

If training one attribute across all layers still takes roughly num_layers × 9 minutes, then that would indicate a real bug, and I would definitely want to take a closer look.

> fixed a bug in batched inference nb with claude code: edit_inter_rep_multi_layers called optimize_one_inter_rep(...) with lr=100, max_epoch=1e5, loss_func=nn.BCELoss(), simplified=True —
           leftover kwargs from an older gradient-descent-based version of the function — but the actual function in these two notebooks only accepts (inter_rep, layer_name, target, probe, N=4, normalized=False)

Thanks again for catching this.

I believe batched_inference.ipynb was mainly created to demonstrate that the intervention can be performed more efficiently in batches when enough GPU memory is available. If your goal is to reproduce the results in the paper, I would recommend running the causality_test_on_*.ipynb notebooks in the same folder instead.

Eyeballing those notebooks, I do not see the same issue there, aside from the sys.path issue you mentioned earlier.

> Used the existing zip files to run the batched inference (lr=100) and causality test for gender (lr=1e-2). See Qn4 below.
The lr should make no difference to the intervened output. The variable that affects the intervention output is N (intervention strength as mentioned in the paper section 5.1)

lr came from an older version of the intervention code that also supported interventions using nonlinear probes. With nonlinear probes, we cannot simply move the hidden representation directly along the probe's weight vector. Instead, we compute the gradient of the probe loss with respect to the intermediate representation and optimize the representation using gradient descent. In that implementation, lr controlled the size of the gradient update.

I am copying the older code down below in case if you are interested:
```python
def optimize_one_inter_rep(inter_rep, layer_name, target, probe,
                           lr=1e-2, max_epoch=128,
                           loss_func=nn.CrossEntropyLoss(),
                           verbose=False, simplified=False, N=4, normalized=False):
    global first_time
    tensor = (inter_rep.clone()).to(torch_device).requires_grad_(True)
    rep_f = lambda: tensor
    target_clone = target.clone().to(torch_device).to(torch.float)

    optimizer = torch.optim.Adam([tensor], lr=lr)

    cur_loss = 1000
    begin_tensor = rep_f().clone().detach()
    cur_input_tensor = rep_f().clone().detach()
    if not simplified:
        if verbose:
            bar = tqdm(range(max_epoch), leave=False)
        else:
            bar = range(max_epoch)
        for i in bar:
            input_tensor = rep_f()
            optimizer.zero_grad()
            probe_seg_out = probe(input_tensor)[0]
            # Compute the loss
            if first_time:
                if logistic:
                    print(probe_seg_out)
                else:
                    print(torch.nn.functional.softmax(probe_seg_out))
                first_time = False
            loss = loss_func(probe_seg_out, target_clone)
            # Call gradient descent
            loss.backward()
            optimizer.step()
            if verbose:
                bar.set_description(f'At layer {layer_name} [{i + 1}/{max_epoch}]; Loss: {loss.item():.3f}')
            if early_stop and abs(cur_input_tensor - input_tensor).mean() < stopping_eta:
                break
            # dist = torch.sqrt(torch.sum((begin_tensor - input_tensor) ** 2))
            dist = torch.sum(torch.sqrt((begin_tensor - input_tensor) ** 2))
            if dist > dist_thresh:
                break
            # cur_loss = loss.item()
            cur_input_tensor = input_tensor.clone().detach()
    else:
        if normalized:
            cur_input_tensor = rep_f() + target_clone.view(1, -1) @ probe.proj[0].weight * N * 100 / rep_f().norm()
        else:
            cur_input_tensor = rep_f() + target_clone.view(1, -1) @ probe.proj[0].weight * N
        dist = torch.sum(torch.sqrt((begin_tensor - cur_input_tensor.clone()) ** 2))
        if layer_name not in distance.keys():
            distance[layer_name] = []
        distance[layer_name].append(dist.item())
        return cur_input_tensor.clone()
   
    dist = torch.sum(torch.sqrt((begin_tensor - input_tensor) ** 2))
    if layer_name not in distance.keys():
        distance[layer_name] = []
    distance[layer_name].append(dist.item())
   
    return rep_f().clone()
```

>Generated a simple [script](https://github.com/rozeappletree/TalkTuner-chatbot-llm-dashboard/blob/main/scripts/cli.py) to replace the web ui, and ran it casually in kaggle T4x2 environment [[kaggle nb here]](https://www.kaggle.com/code/l0new0lf/mats-talk-tuner), output seemed decent enough, I need to make it better. See Qn5.

Very cool! One thing you could try is to play around with the intervention strength N. 

>Where is the code using which you have generated the projection as in the paper shown below?

That seems to be a plot from the appendices. I will see if its code is still on my cluster. To generate that plot, you just need the last token hidden states collected from the conversations in the age dataset. Then, project them on the reading probe's weights and use them as the x-coordinates. Finally, find the null space of the reading probe's vector and project the activations on one basis vector of the null space. You should be able to find many basis vectors which explains a lot of additional variance in the activations but align more closely with the control probe's weight vector.

>Can I use the code in [batched inference nb](https://github.com/rozeappletree/TalkTuner-chatbot-llm-dashboard/blob/main/notebooks/causality_tests/batched_inference.ipynb) to recreate web ui? or do I need extra code?

I recommend checking Cat McGee's implementation first and you can wait for me to clean up our server code, which should be done soon, too. https://github.com/catmcgee/TalkTuner/tree/main/dashboard

>I am assuming that in the causality tests notebooks, in order to use the reading probe classifier, I just need to replace the control probe checkpoint path to reading probe checkpoints. Correct me if it is not that easy! I am also looking for notebooks used for comparing the probes.

That's right."

**Comment by rozeappletree** [comment]
Thanks @yc015 for putting in time and effort for the detailed response. really appreciated! I see that most of the pieces are already available with some (notably?) insignificant pieces missing. 

Will let you know if I find any issues.

**Comment by yc015** [comment]
Awesome! I will work on open-sourcing the web app code over the next few days. In the meantime, if you notice any other bugs, feel free to share them here and I will take a look together.

**Comment by rozeappletree** [comment]
> In Section 4.1, we did mention that both GPT-3.5 and LLaMA2-Chat were used to synthesize the data: “Using the role-playing technique, we generated synthetic conversations using GPT-3.5 and LLaMa2Chat.”

Ah. I took the code-first approach to understand the repo. 

Both the notebooks inside [data generation folder](https://github.com/yc015/TalkTuner-chatbot-llm-dashboard/tree/main/notebooks/dataset_generation)  generate synthetic conversations independently, with a single model self-playing both sides in one completion. 

Missed this part, thanks for letting me know!

**Comment by yc015** [comment]
In case you were wondering, I haven’t forgotten about it and I am still working on it.

**Comment by rozeappletree** [comment]
Thanks for letting me know @yc015 

I'm to write something cool too, will share it here as well!

<img width="3000" height="4000" alt="Image" src="https://github.com/user-attachments/assets/97fd8bee-1c4c-4e71-be1f-07e685e0853c" />

Best,
R

**Comment by yc015** [comment]
Awesome. Looking forward to it!

In the meanwhile, I have uploaded the Web App code in https://github.com/yc015/TalkTuner-chatbot-llm-dashboard/tree/main/dashboard_v1. 

It's a bit complicate to serve the App, so please follow the README closely. Let me know if you encountered any difficulties, I will keep this issue open for a short while in case you need any additional help!

**Comment by rozeappletree** [comment]
Thanks @yc015, was moving at faster pace as today is the deadline for MATS, I one-shotted it with claude code + opus high & I got exactly what I needed back then. 

my one shotted code: https://github.com/rozeappletree/mats12/blob/main/webui/app.py

Will crossverify!

**Comment by yc015** [comment]
Cool! Good luck with your application

**Comment by rozeappletree** [comment]
thx! in case you are interested, here is the proj: https://github.com/rozeappletree/SeeGULL

(was supposed to write it better, but on a high level, the readme covers it)

**Comment by yc015** [comment]
Looks great! I would appreciate if you could credit / cite this repo if it was helpful in building your project.

**Comment by rozeappletree** [comment]
> Looks great! I would appreciate if you could credit / cite this repo if it was helpful in building your project.

Noted! Will add to the readme with the release of V1 !
